### PR TITLE
 refactor(client): remove manual client arc counting 

### PR DESCRIPTION
**Pull request by dpc**
On top of #4482 

Use two levels of `Arc` instead. While at it, rename and document things
better.

### CODE REVIEW

File: fedimint-client/src/lib.rs [hunk 1]
```diff
@@ -594,8 +608,8 @@ pub(crate) struct ClientWeak {
 }
 
 impl ClientWeak {
-    pub fn upgrade(&self) -> Option<ClientHandle> {
-        Weak::upgrade(&self.inner).map(ClientHandle::new)
+    pub fn upgrade(&self) -> Option<ClientStrong> {
```
Comment (elsirion) [inline fedimint-client/src/lib.rs:611]: Ah, I think you fixed that problem here

**Review by elsirion — APPROVED** [review]
I think in total the PR is good, just one intermediately broken commit.

**Comment by dpc** [comment]
> I think in total the PR is good, just one intermediately broken commit.

Fixed at first, but then decided to squash the 3 last commits, as it doesn't help anything to have a commit with a problem.

**Review by elsirion — APPROVED** [review]

**Review by Kodylow — APPROVED** [review]
tACK

**Pull request merged**